// @okxweb3/coin-base/dist/currency.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CoinCurveMap = exports.MultiAddressCoins = exports.segwitType = exports.Currency = void 0;
var Currency;
(function (Currency) {
    Currency[Currency["NULL"] = -1] = "NULL";
    Currency[Currency["BTC"] = 0] = "BTC";
    Currency[Currency["TBTC"] = 1] = "TBTC";
    Currency[Currency["ETH"] = 60] = "ETH";
    Currency[Currency["LTC"] = 2] = "LTC";
    Currency[Currency["DOGE"] = 3] = "DOGE";
    Currency[Currency["BCH"] = 145] = "BCH";
    Currency[Currency["BSV"] = 236] = "BSV";
    Currency[Currency["TRX"] = 195] = "TRX";
    Currency[Currency["SOL"] = 501] = "SOL";
    Currency[Currency["APTOS"] = 637] = "APTOS";
    Currency[Currency["OMNI_USDT"] = 20001] = "OMNI_USDT";
    Currency[Currency["OMNI_USDT_TEST"] = 20002] = "OMNI_USDT_TEST";
    Currency[Currency["SUI"] = 784] = "SUI";
    Currency[Currency["ATOM"] = 118] = "ATOM";
    Currency[Currency["OSMO"] = 1000] = "OSMO";
    Currency[Currency["EVMOS"] = 1001] = "EVMOS";
    Currency[Currency["AXL"] = 1002] = "AXL";
    Currency[Currency["CRO"] = 1003] = "CRO";
    Currency[Currency["Iris"] = 1004] = "Iris";
    Currency[Currency["Juno"] = 1005] = "Juno";
    Currency[Currency["Kava"] = 1006] = "Kava";
    Currency[Currency["Kujira"] = 1007] = "Kujira";
    Currency[Currency["SCRT"] = 1008] = "SCRT";
    Currency[Currency["Stargaze"] = 1009] = "Stargaze";
    Currency[Currency["Terra"] = 1010] = "Terra";
    Currency[Currency["ZKSPACE"] = 1011] = "ZKSPACE";
    Currency[Currency["ZKSYNC"] = 804] = "ZKSYNC";
    Currency[Currency["Stx"] = 5757] = "Stx";
    Currency[Currency["SEI"] = 2837] = "SEI";
    Currency[Currency["WAX"] = 14001] = "WAX";
    Currency[Currency["Starknet"] = 9004] = "Starknet";
    Currency[Currency["ADA"] = 1815] = "ADA";
    Currency[Currency["INJ"] = 2892] = "INJ";
    Currency[Currency["Celestia"] = 2854] = "Celestia";
    Currency[Currency["DYDX"] = 2897] = "DYDX";
    Currency[Currency["Kaspa"] = 111111] = "Kaspa";
    Currency[Currency["NostrAssets"] = 1237] = "NostrAssets";
})(Currency = exports.Currency || (exports.Currency = {}));
var segwitType;
(function (segwitType) {
    segwitType[segwitType["SEGWIT_NESTED"] = 1] = "SEGWIT_NESTED";
    segwitType[segwitType["SEGWIT_NESTED_49"] = 2] = "SEGWIT_NESTED_49";
    segwitType[segwitType["SEGWIT_NATIVE"] = 3] = "SEGWIT_NATIVE";
    segwitType[segwitType["SEGWIT_TAPROOT"] = 4] = "SEGWIT_TAPROOT";
})(segwitType = exports.segwitType || (exports.segwitType = {}));
exports.MultiAddressCoins = [
    Currency.BTC, Currency.TBTC, Currency.LTC, Currency.DOGE, Currency.BSV, Currency.OMNI_USDT, Currency.OMNI_USDT_TEST
];
const secp256k1 = "secp256k1";
const ed25519 = "ed25519";
exports.CoinCurveMap = new Map([
    [Currency.BTC, secp256k1],
    [Currency.TBTC, secp256k1],
    [Currency.ETH, secp256k1],
    [Currency.LTC, secp256k1],
    [Currency.DOGE, secp256k1],
    [Currency.BCH, secp256k1],
    [Currency.BSV, secp256k1],
    [Currency.TRX, secp256k1],
    [Currency.OMNI_USDT, secp256k1],
    [Currency.OMNI_USDT_TEST, secp256k1],
    [Currency.ATOM, secp256k1],
    [Currency.OSMO, secp256k1],
    [Currency.EVMOS, secp256k1],
    [Currency.AXL, secp256k1],
    [Currency.CRO, secp256k1],
    [Currency.Iris, secp256k1],
    [Currency.Juno, secp256k1],
    [Currency.Kava, secp256k1],
    [Currency.Kujira, secp256k1],
    [Currency.SCRT, secp256k1],
    [Currency.Stargaze, secp256k1],
    [Currency.Terra, secp256k1],
    [Currency.SEI, secp256k1],
    [Currency.Celestia, secp256k1],
    [Currency.DYDX, secp256k1],
    [Currency.SOL, ed25519],
    [Currency.APTOS, ed25519],
    [Currency.SUI, ed25519],
]);
