// @okxweb3/coin-base/dist/error.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EstimateFeeError = exports.validSignedTransactionError = exports.GetPayLoadError = exports.HardwareDerivePubKeyError = exports.CalcTxHashError = exports.SignMsgStep3Error = exports.SignMsgSetupError = exports.Ed25519Step3Error = exports.Ed25519Step2Error = exports.Ed25519Step1Error = exports.Ed25519SetupError = exports.EcdsaStep3Error = exports.EcdsaStep2Error = exports.EcdsaStep1Error = exports.EcdsaSetupError = exports.MPCShakeHandsError = exports.MPCStepError = exports.MPCAesKeyMissError = exports.GetHardWareMessageHashError = exports.GetHardwareSignedTransactionError = exports.GetHardwareRawTransactionError = exports.GetMpcTransactionError = exports.GetMpcRawTransactionError = exports.DerivePathError = exports.GenMnemonicError = exports.GenPrivateKeyError = exports.UnsupportedCoinError = exports.SignMsgError = exports.SignTxError = exports.ValidAddressError = exports.NewAddressError = exports.NotImplementedError = void 0;
exports.NotImplementedError = "no implementation method";
exports.NewAddressError = "generate address error";
exports.ValidAddressError = "valid address error";
exports.SignTxError = "sign tx error";
exports.SignMsgError = "sign message error";
exports.UnsupportedCoinError = "unsupported currency";
exports.GenPrivateKeyError = "generate private key error";
exports.GenMnemonicError = "generate mnemonic error";
exports.DerivePathError = "derive hdPath error";
exports.GetMpcRawTransactionError = "get mpc raw transaction error";
exports.GetMpcTransactionError = "get mpc transaction error";
exports.GetHardwareRawTransactionError = "get hardware raw transaction error";
exports.GetHardwareSignedTransactionError = "get hardware signed transaction error";
exports.GetHardWareMessageHashError = "get hardware message hash error";
exports.MPCAesKeyMissError = "mpc get aes key error";
exports.MPCStepError = "mpc step error";
exports.MPCShakeHandsError = 'mpc shake hands error';
exports.EcdsaSetupError = 'ecdsa setup error';
exports.EcdsaStep1Error = 'ecdsa step1 error';
exports.EcdsaStep2Error = 'ecdsa step2 error';
exports.EcdsaStep3Error = 'ecdsa step3 error';
exports.Ed25519SetupError = 'ed25519 setup error';
exports.Ed25519Step1Error = 'ed25519 step1 error';
exports.Ed25519Step2Error = 'ed25519 step2 error';
exports.Ed25519Step3Error = 'ed25519 step3 error';
exports.SignMsgSetupError = 'sign message setup error';
exports.SignMsgStep3Error = 'sign message step3 error';
exports.CalcTxHashError = "calculate tx hash error";
exports.HardwareDerivePubKeyError = "derive sub pub key error";
exports.GetPayLoadError = "get payload error";
exports.validSignedTransactionError = "valid signed transaction error";
exports.EstimateFeeError = "estimate fee error";
