// @okxweb3/coin-bitcoin/dist/wallet/LtcWallet.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LtcWallet = exports.litecoin = void 0;
const coin_base_1 = require("@okxweb3/coin-base");
const BtcWallet_1 = require("./BtcWallet");
exports.litecoin = {
    messagePrefix: '\x19Litecoin Signed Message:\n',
    bech32: 'ltc',
    bip32: {
        public: 0x019da462,
        private: 0x019d9cfe,
    },
    pubKeyHash: 0x30,
    scriptHash: 0x32,
    wif: 0xb0,
};
class LtcWallet extends BtcWallet_1.BtcWallet {
    network() {
        return exports.litecoin;
    }
    async getDerivedPath(param) {
        if (!param.segwitType) {
            return `m/44'/2'/0'/0/${param.index}`;
        }
        if (param.segwitType == coin_base_1.segwitType.SEGWIT_NESTED) {
            return `m/84'/2'/0'/0/${param.index}`;
        }
        else if (param.segwitType == coin_base_1.segwitType.SEGWIT_NESTED_49) {
            return `m/49'/2'/0'/0/${param.index}`;
        }
        else if (param.segwitType == coin_base_1.segwitType.SEGWIT_NATIVE) {
            return `m/84'/2'/0'/0/${param.index}`;
        }
        else {
            return Promise.reject(coin_base_1.DerivePathError);
        }
    }
}
exports.LtcWallet = LtcWallet;
