// @okxweb3/coin-bitcoin/dist/rune.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildRuneData = exports.fromVarInt = exports.encodeToVec = exports.toVarInt = void 0;
const bscript = __importStar(require("./bitcoinjs-lib/script"));
const ops_1 = require("./bitcoinjs-lib/ops");
function encode(n) {
    let payload = [];
    encodeToVec(n, payload);
    return new Uint8Array(payload);
}
exports.toVarInt = encode;
function encodeToVec(n, payload) {
    let i = 18;
    const out = new Array(19).fill(0);
    out[i] = Number(n & BigInt(0x7F));
    while (n > BigInt(0x7F)) {
        n = n / BigInt(128) - BigInt(1);
        i--;
        out[i] = Number(n & BigInt(0xFF)) | 0x80;
    }
    payload.push(...out.slice(i));
}
exports.encodeToVec = encodeToVec;
function decode(buffer) {
    let n = BigInt(0);
    let i = 0;
    while (true) {
        const b = BigInt(buffer[i]);
        if (b < BigInt(128)) {
            return [n + b, i + 1];
        }
        n += b - BigInt(127);
        n = n * BigInt(128);
        i++;
        if (i >= buffer.length) {
            throw new Error("Varint decoding error: buffer overflow");
        }
    }
}
exports.fromVarInt = decode;
const TAG_BODY = BigInt(0);
function buildRuneData(isMainnet, edicts) {
    let payload = [];
    if (edicts.length > 0) {
        encodeToVec(TAG_BODY, payload);
        edicts.sort((a, b) => a.id - b.id);
        let id = 0;
        for (const edict of edicts) {
            encodeToVec(BigInt(edict.id - id), payload);
            encodeToVec(BigInt(edict.amount), payload);
            encodeToVec(BigInt(edict.output), payload);
            id = edict.id;
        }
    }
    let prefix;
    if (isMainnet) {
        prefix = 'R';
    }
    else {
        prefix = 'RUNE_TEST';
    }
    const opReturnScript = bscript.compile([ops_1.OPS.OP_RETURN, Buffer.from(prefix), Buffer.from(payload)]);
    return opReturnScript;
}
exports.buildRuneData = buildRuneData;
